### ISSUE TITLE
ZIP32 implementation in PHP

### ISSUE THREAD

**Issue by luke424** [zcash/zcash#3894]
I sent in an email about this and was asked to put it here...

Is there any plans to write a ZIP32 implementation in PHP? I'd love to take Sapling payments but realistically I'd need such to be done in PHP and setting up a bunch of zcashd instances is unrealistic.

This should do things like: create a master key, derive child keys, derive payment addresses given an index and public master key, derive the minimal information needed to check for a payment (but not spend) given the equivalent of a view master key, and derive the information needed to spend coins given a private master key.

**Comment by daira** [comment]
The most realistic approach is probably writing a binding to librustzcash in PHP. It behaves like any C library in terms of FFI compatibility.

**Comment by luke424** [comment]
Why couldn't it be done in PHP though? Or even I suppose pure Python is fine, as these are both languages supported on most web hosting environments. I've never heard of Rust before zcash and don't think most web hosting admins are familiar with it either.

A lot of the instances of our software have to run on servers where it's really unrealistic for us to have highly specific dependencies installed. I know for example the Bitwasp PHP implementation for Bitcoin was pretty big in terms of Bitcoin being taken on more sites since it made it much easier for web admins to set up without relying on third parties.

**Comment by daira** [comment]
PHP is not a suitable language in which to write secure cryptographic code.

You don't need to know anything about Rust to use librustzcash. It provides a C-compatible API, which is [entirely feasible to use from PHP](https://stackoverflow.com/questions/2479402/calling-c-c-library-function-from-php).

**Comment by luke424** [comment]
Alright understood.

To further - specifically the thing I personally wanted to do in pure PHP was just to take a public master key and derive payment addresses from it. 

The remaining (checking payments and sending) can be done as back office operations, which are also the sensitive keys (private view and spend key). We don't mind dependencies for this as this is done on a single machine.

However, I don't see why generating the addresses has to be done in a particularly "hardened" manner. The most an attacker could possibly get out of it is the master public key, and the most they could do with it is generate payment addresses, but they wouldn't even be able to check which addresses have been used given the nature of a confidential transaction (unless I misunderstood how this works).

**Comment by daira** [comment]
ZIP 32 doesn't support deriving a child payment address from a parent payment address (the cryptography does not allow this, [unlike in Bitcoin](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#Public_parent_key_rarr_public_child_key)).

If you mean [deriving a child extended viewing key from a parent extended viewing key](https://github.com/zcash/zips/blob/master/zip-0032.rst#deriving-a-child-extended-full-viewing-key), that still requires the code to handle secrets, even if they are viewing keys rather than spending keys.

**Comment by luke424** [comment]
So I suppose I have a fundamental misunderstanding of how this works.

We had one set up for Monero, where we use this: https://github.com/monero-integrations/monerophp/blob/master/src/subaddress.php It is accepted that in a worst case scenario that an attacker would be able to recover our secret viewing key (which we protect against by switching them out regularly). As I understand of crypto one of PHP's big weaknesses is not able to generate secure randomness or otherwise leaking data but ideally we'd want our addresses deterministic. The random master key is generated on our back office and we can set up a single VM to run the rust plugin on there easily.

My ideal objective is this:

Public servers handling payments will be given some master token from which it can derive unique Sapling payment addresses that we can give to users.

You shouldn't be able to link multiple payments together just by looking at the blockchain, and given a large list of addresses, you shouldn't be able to pick out which ones are ours. With a Monero integrated address, it can do the first thing (the address on the blockchain itself is derived and single-use), but it doesn't do the second (you could decode the integrated address and just look for which addresses have the same public spend/view keys).

So on zcash then, would I be able to use the same child payment address just with different diversifiers? Would this produce valid addresses? Would someone be able to find the common "root payment address" without the diversifier given such address? And... any documentation so we could do this in just PHP?

I don't know if this made sense, I'm new to Zcash terminology. Sorry for being confusing, I find this really interesting and would like to learn/understand it fully and so I can implement this the best way possible.

**Comment by daira** [comment]
There's no problem in the abstract in providing a PHP API to derive a child extended viewing key from a parent extended viewing key. However, since PHP isn't suited to writing secure cryptographic code, I strongly recommend that you do it by calling into librustzcash. We won't be writing code in PHP.

**Comment by luke424** [comment]
Understood.

Is it possible then to suggest that zip32 include a section like this: https://en.bitcoin.it/wiki/Technical_background_of_version_1_Bitcoin_addresses "How to create Bitcoin Address" section?

Basically step by step description of how to go from the initial random data, down to an address that can be paid to. We did go through the zcash protocol pdf file and the current zip32 github page and it was a little confusing.

**Comment by daira** [comment]
Documentation request filed as #3931. Is that sufficient to resolve this ticket?

**Comment by ioptio** [comment]
Let's add this to the encouraged community work instead of closing it.

**Comment by daira** [comment]
I don't think we should be encouraging anyone in the community to write cryptographic code in PHP.
